src/BTVNanoCommissioning/utils/correction.py

```diff
@@ -301,173 +345,956 @@ def add_jec_variables(jets, event_rho):
+                            jet_eta,
+                            jet_pt,
+                            "down",
+                            correct_map["JMAR_cfg"][sf],
+                        ),
                     )
+                    sfs_all_up = sfs_up * sfs_all_up
+                    sfs_all_down = sfs_down * sfs_all_down
+            sfs_all = sfs * sfs_all
+        if syst:
+            weights.add(sf, sfs_all, sfs_all_up, sfs_all_down)
+        else:
+            weights.add(sf, sfs_all)
+
+
+def add_pdf_weight(weights, pdf_weights):
+    nom = np.ones(len(weights.weight()))
+    up = np.ones(len(weights.weight()))
+    down = np.ones(len(weights.weight()))
+
+    # NNPDF31_nnlo_hessian_pdfas
+    # https://lhapdfsets.web.cern.ch/current/NNPDF31_nnlo_hessian_pdfas/NNPDF31_nnlo_hessian_pdfas.info
+    if pdf_weights is not None and "306000 - 306102" in pdf_weights.__doc__:
+        # Hessian PDF weights
+        # Eq. 21 of https://arxiv.org/pdf/1510.03865v1.pdf
+        arg = pdf_weights[:, 1:-2] - np.ones((len(weights.weight()), 100))
+        summed = ak.sum(np.square(arg), axis=1)
+        pdf_unc = np.sqrt((1.0 / 99.0) * summed)
+        weights.add("PDF_weight", nom, pdf_unc + nom)
 
-            weight = weight * sfs
-        elif "correctionlib" in str(type(correct_map["MUO"])):
-            weight = weight * correct_map["MUO"][correct_map["MUO_cfg"][sf]].evaluate(
-                sf[sf.find(" ") + 1 :], mu_eta, mu_pt, "sf"
+        # alpha_S weights
+        # Eq. 27 of same ref
+        as_unc = 0.5 * (pdf_weights[:, 102] - pdf_weights[:, 101])
+        weights.add("aS_weight", nom, as_unc + nom)
+
+        # PDF + alpha_S weights
+        # Eq. 28 of same ref
+        pdfas_unc = np.sqrt(np.square(pdf_unc) + np.square(as_unc))
+        weights.add("PDFaS_weight", nom, pdfas_unc + nom)
+
+    else:
+        weights.add("aS_weight", nom, up, down)
+        weights.add("PDF_weight", nom, up, down)
+        weights.add("PDFaS_weight", nom, up, down)
+
+
+# https://twiki.cern.ch/twiki/bin/viewauth/CMS/TopPtReweighting#TOP_PAG_corrections_based_on_the
+def top_pT_sf_formula(pt):
+    return 0.103 * np.exp(-0.0118 * pt) - 0.000134 * pt + 0.973
+
+
+def top_pT_reweighting(gen):
+    #     """
+    #     Apply this SF only to TTbar datasets! Updated to latest suggestion
+    #     Documentation:
+    #         - https://twiki.cern.ch/twiki/bin/viewauth/CMS/TopPtReweighting
+    #         - https://twiki.cern.ch/twiki/bin/viewauth/CMS/TopPtReweighting#TOP_PAG_corrections_based_on_the
+    #     """
+    top = gen[(gen.pdgId == 6) & gen.hasFlags(["isLastCopy"])]
+    anti_top = gen[(gen.pdgId == -6) & gen.hasFlags(["isLastCopy"])]
+    return np.sqrt(
+        top_pT_sf_formula(ak.flatten(top.pt, axis=-1))
+        * top_pT_sf_formula(ak.flatten(anti_top.pt, axis=-1))
+    )
+
+
+# Jennet adds PS weights
+# https://github.com/andrzejnovak/boostedhiggs/blob/master/boostedhiggs/corrections.py#L88-L108
```

Comment:
please add also to the description of this PR, which parts of correction.py have been taken from the external source

Reply:
Thanks for review the PR, I updated the comments now :)